File: internal/agent/comm/loopback.go
package comm

import (
	"context"
	"fmt"
	"sync"
)

// LoopbackPlugin is an in-memory comm plugin for testing and development.
// It delivers sent messages back to the handler, enabling local testing
// of the full comm pipeline without external infrastructure.
type LoopbackPlugin struct {
	handler   func(CommMessage)
	connected bool
	topics    map[string]bool
	agentID   string
	mu        sync.RWMutex
}

// NewLoopbackPlugin creates a new loopback plugin
func NewLoopbackPlugin() *LoopbackPlugin {
	return &LoopbackPlugin{
		topics: make(map[string]bool),
	}
}

func (p *LoopbackPlugin) Name() string    { return "loopback" }
func (p *LoopbackPlugin) Version() string { return "1.0.0" }

func (p *LoopbackPlugin) Connect(ctx context.Context, config map[string]string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.connected = true
	fmt.Printf("[Comm:loopback] Connected\n")
	return nil
}

func (p *LoopbackPlugin) Disconnect(ctx context.Context) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.connected = false
	p.topics = make(map[string]bool)
	fmt.Printf("[Comm:loopback] Disconnected\n")
	return nil
}

func (p *LoopbackPlugin) IsConnected() bool {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.connected
}

func (p *LoopbackPlugin) Send(ctx context.Context, msg CommMessage) error {
	p.mu.RLock()
	connected := p.connected
	p.mu.RUnlock()

	if !connected {
		return fmt.Errorf("loopback plugin not connected")
	}

	fmt.Printf("[Comm:loopback] Message sent: from=%s to=%s topic=%s\n", msg.From, msg.To, msg.Topic)
	return nil
}

func (p *LoopbackPlugin) Subscribe(ctx context.Context, topic string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if !p.connected {
		return fmt.Errorf("loopback plugin not connected")
	}
	p.topics[topic] = true
	fmt.Printf("[Comm:loopback] Subscribed to: %s\n", topic)
	return nil
}

func (p *LoopbackPlugin) Unsubscribe(ctx context.Context, topic string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if !p.connected {
		return fmt.Errorf("loopback plugin not connected")
	}
	delete(p.topics, topic)
	fmt.Printf("[Comm:loopback] Unsubscribed from: %s\n", topic)
	return nil
}

func (p *LoopbackPlugin) Register(ctx context.Context, agentID string, card *AgentCard) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.agentID = agentID
	if card != nil {
		fmt.Printf("[Comm:loopback] Registered agent: %s skills=%d\n", agentID, len(card.Skills))
	} else {
		fmt.Printf("[Comm:loopback] Registered agent: %s\n", agentID)
	}
	return nil
}

func (p *LoopbackPlugin) Deregister(ctx context.Context) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.agentID = ""
	return nil
}

func (p *LoopbackPlugin) SetMessageHandler(handler func(msg CommMessage)) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.handler = handler
}

// InjectMessage simulates receiving a message from the network.
// Used for testing — delivers the message to the handler as if it arrived externally.
func (p *LoopbackPlugin) InjectMessage(msg CommMessage) {
	p.mu.RLock()
	handler := p.handler
	subscribed := p.topics[msg.Topic]
	p.mu.RUnlock()

	if handler == nil {
		fmt.Printf("[Comm:loopback] No handler set, dropping message\n")
		return
	}
	if !subscribed {
		fmt.Printf("[Comm:loopback] Not subscribed to topic %q, dropping message\n", msg.Topic)
		return
	}
	handler(msg)
}

File: crates/comm/src/loopback.rs
use std::collections::{HashMap, HashSet};
use std::sync::RwLock;

use tracing::{debug, info, warn};

use crate::{AgentCard, CommError, CommMessage, CommPlugin, MessageHandler};

/// In-memory comm plugin for testing. Delivers sent messages back to the handler.
pub struct LoopbackPlugin {
    inner: RwLock<Inner>,
}

struct Inner {
    handler: Option<MessageHandler>,
    connected: bool,
    topics: HashSet<String>,
    agent_id: String,
}

impl LoopbackPlugin {
    pub fn new() -> Self {
        Self {
            inner: RwLock::new(Inner {
                handler: None,
                connected: false,
                topics: HashSet::new(),
                agent_id: String::new(),
            }),
        }
    }

    /// Simulate receiving a message from the network (for testing).
    pub fn inject_message(&self, msg: CommMessage) {
        let inner = self.inner.read().unwrap();
        let handler = inner.handler.clone();
        let subscribed = inner.topics.contains(&msg.topic);
        drop(inner);

        match handler {
            None => warn!("loopback: no handler set, dropping message"),
            Some(h) if !subscribed => {
                warn!(topic = msg.topic.as_str(), "loopback: not subscribed to topic, dropping message");
            }
            Some(h) => h(msg),
        }
    }
}

impl Default for LoopbackPlugin {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait::async_trait]
impl CommPlugin for LoopbackPlugin {
    fn name(&self) -> &str {
        "loopback"
    }
    fn version(&self) -> &str {
        "1.0.0"
    }

    async fn connect(&self, _config: HashMap<String, String>) -> Result<(), CommError> {
        let mut inner = self.inner.write().unwrap();
        inner.connected = true;
        info!("loopback connected");
        Ok(())
    }

    async fn disconnect(&self) -> Result<(), CommError> {
        let mut inner = self.inner.write().unwrap();
        inner.connected = false;
        inner.topics.clear();
        info!("loopback disconnected");
        Ok(())
    }

    fn is_connected(&self) -> bool {
        self.inner.read().unwrap().connected
    }

    async fn send(&self, msg: CommMessage) -> Result<(), CommError> {
        let inner = self.inner.read().unwrap();
        if !inner.connected {
            return Err(CommError::NotConnected);
        }
        debug!(from = msg.from.as_str(), to = msg.to.as_str(), topic = msg.topic.as_str(), "loopback message sent");
        Ok(())
    }

    async fn subscribe(&self, topic: &str) -> Result<(), CommError> {
        let mut inner = self.inner.write().unwrap();
        if !inner.connected {
            return Err(CommError::NotConnected);
        }
        inner.topics.insert(topic.to_string());
        debug!(topic, "loopback subscribed");
        Ok(())
    }

    async fn unsubscribe(&self, topic: &str) -> Result<(), CommError> {
        let mut inner = self.inner.write().unwrap();
        if !inner.connected {
            return Err(CommError::NotConnected);
        }
        inner.topics.remove(topic);
        debug!(topic, "loopback unsubscribed");
        Ok(())
    }

    async fn register(&self, agent_id: &str, card: &AgentCard) -> Result<(), CommError> {
        let mut inner = self.inner.write().unwrap();
        inner.agent_id = agent_id.to_string();
        info!(agent = agent_id, skills = card.skills.len(), "loopback registered agent");
        Ok(())
    }

    async fn deregister(&self) -> Result<(), CommError> {
        let mut inner = self.inner.write().unwrap();
        inner.agent_id.clear();
        Ok(())
    }

    fn set_message_handler(&self, handler: MessageHandler) {
        let mut inner = self.inner.write().unwrap();
        inner.handler = Some(handler);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Arc;
    use std::sync::atomic::{AtomicBool, Ordering};

    #[tokio::test]
    async fn test_loopback_lifecycle() {
        let plugin = LoopbackPlugin::new();
        assert!(!plugin.is_connected());

        plugin.connect(HashMap::new()).await.unwrap();
        assert!(plugin.is_connected());

        plugin.subscribe("test-topic").await.unwrap();
        plugin.disconnect().await.unwrap();
        assert!(!plugin.is_connected());
    }

    #[tokio::test]
    async fn test_loopback_inject() {
        let plugin = LoopbackPlugin::new();
        let received = Arc::new(AtomicBool::new(false));
        let received2 = received.clone();

        plugin.connect(HashMap::new()).await.unwrap();
        plugin.subscribe("test").await.unwrap();
        plugin.set_message_handler(Arc::new(move |_msg| {
            received2.store(true, Ordering::SeqCst);
        }));

        plugin.inject_message(CommMessage {
            id: "1".into(),
            from: "a".into(),
            to: "b".into(),
            topic: "test".into(),
            conversation_id: String::new(),
            msg_type: crate::CommMessageType::Message,
            content: "hello".into(),
            metadata: HashMap::new(),
            timestamp: 0,
            human_injected: false,
            human_id: None,
            task_id: None,
            correlation_id: None,
            task_status: None,
            artifacts: vec![],
            error: None,
        });

        assert!(received.load(Ordering::SeqCst));
    }
}
